File: 2025/1/puzzle.exs
IO.puts("Day 1: Historian Hysteria")

defmodule Helpers do
  def sortAtIndex(list, index) do
    list
    |> Enum.map(&String.to_integer(Enum.at(&1, index, "0")))
    |> Enum.sort()
  end
end

case File.read("input.txt") do
  {:ok, contents} ->
    nums =
      contents
      |> String.split("\n", trim: true)
      |> Enum.map(&String.split(&1, "   "))

    lOne = Helpers.sortAtIndex(nums, 0)
    lTwo = Helpers.sortAtIndex(nums, 1)

    Enum.zip(lOne, lTwo)
    |> Enum.map(fn {a, b} -> abs(a - b) end)
    |> Enum.sum()
    |> IO.inspect(label: "Part 1")

    frequencies = Enum.frequencies(lTwo)

    lOne
    |> Enum.map(&(&1 * Map.get(frequencies, &1, 0)))
    |> Enum.sum()
    |> IO.inspect(label: "Part 2")

  {:error, :enoent} ->
    nil
end

File: 2024/1/puzzle.py
print("Day 1: Historian Hysteria")

with open("input.txt", encoding="utf-8") as f:
    lines = [x.replace("\n", "") for x in list(f)]

total = 0
lOne = []
lTwo = []

for line in lines:
    nums = line.split("   ")
    lOne.append(int(nums[0]))
    lTwo.append(int(nums[1]))

lOne.sort()
lTwo.sort()

for i, x in enumerate(lOne):
    total += abs(lOne[i] - lTwo[i])

print("Part 1:", total)

simScores = []

for x in lOne:
    simScores.append(x * lTwo.count(x))

print("Part 2:", sum(simScores))
